### PR TITLE
VIMC-3617: Add support for redirecting orderly logs to a file

### PR DESCRIPTION
**Pull request by r-ash**
This PR will
* Add support for `orderly_run` and `orderly_commit` to write logs to a `orderly.log` file based on a config flag
* Use config flag in CLI run

### CODE REVIEW

File: R/recipe_commit.R [hunk 1]
```diff
@@ -37,7 +37,11 @@ orderly_commit <- function(id, name = NULL, root = NULL, locate = TRUE) {
     }
   }
   workdir <- file.path(path_draft(config$root), name, id)
-  recipe_commit(workdir, config$root)
+  
+  capture <- inherits(root, "orderly_config") &&
```
Comment (richfitz) [inline R/recipe_commit.R:41]: why use `root` here and not `config`?  should just be able to write:

```
capture_log <- isTRUE(config$get_run_option("capture_log"))
```

I think.

File: R/recipe_commit.R [hunk 2]
```diff
@@ -37,7 +37,11 @@ orderly_commit <- function(id, name = NULL, root = NULL, locate = TRUE) {
     }
   }
   workdir <- file.path(path_draft(config$root), name, id)
-  recipe_commit(workdir, config$root)
+  
+  capture <- inherits(root, "orderly_config") && 
+    isTRUE(root$get_run_option("capture_log"))
+  logfile <- file.path(path_draft(config$root), name, id, "orderly.log")
+  conditional_capture_log(capture, logfile, recipe_commit(workdir, config$root))
```
Comment (richfitz) [inline R/recipe_commit.R:44]: I think this will overwrite the actual run log?

File: R/recipe_run.R [hunk 3]
```diff
@@ -143,17 +143,30 @@ orderly_run <- function(name = NULL, parameters = NULL, envir = NULL,
...
+    withr::with_envvar(
+      orderly_envir_read(config$root),
+      recipe_run(info, parameters, envir, config, echo = echo,
+                 instance = instance))
+  }
+  
+  capture <- inherits(root, "orderly_config") && 
+    isTRUE(root$get_run_option("capture_log"))
+  logfile <- tempfile()
+  info <- conditional_capture_log(capture, logfile, run(capture))
```
Comment (richfitz) [inline R/recipe_run.R:169]: I had wondered about writing it in a withr style:

```
info <- conditional_capture_log(capture, logfile, {
  info <- recipe_prepare(config, name, id_file, ref, fetch, message,
                         use_draft, parameters, remote, tags = tags,
                         batch_id = batch_id)

  if (capture) {
    dest <- path_draft(config$root)
    on.exit(file_copy(logfile, file.path(dest, name, info$id, "orderly.log")), 
            add = TRUE)
  }

  recipe_current_run_set(info)
  on.exit(recipe_current_run_clear(), add = TRUE)

  withr::with_envvar(
    orderly_envir_read(config$root),
    recipe_run(info, parameters, envir, config, echo = echo,
               instance = instance))
})
```

what do you think?

File: R/util.R [hunk 4]
```diff
@@ -196,7 +196,12 @@ pasteq <- function(x, sep = ", ") {
 
 capture_log <- function(expr, filename) {
   ## nolint start
-  con <- file(filename, "w")
+  if (file.exists(filename)) {
+    mode = "a"
```
Comment (richfitz) [inline R/util.R:200]: why not `<-`? and why is nolint required?

Comment (richfitz) [reply R/util.R:200]: you can also just write 

```
mode <- if (file.exists(filename)) "a" else "w"
```

File: R/main.R [hunk 5]
```diff
@@ -136,38 +136,27 @@ main_do_run <- function(x) {
...
   } else {
-    log <- tempfile()
-    ## we should run this with try() so that we can capture logs there
-    id <- capture_log(main_run(), log)
-    dest <- (if (commit) path_archive else path_draft)(config$root)
-    file_copy(log, file.path(dest, name, id, "orderly.log"))
+    config$add_run_option("capture_log", FALSE)
   }
 
+  if (pull) {
```
Comment (richfitz) [inline R/main.R:147]: we'll have to sort this mess out at some point too - seems bizarre that we pull here but fetch within the run command? What was I thinking?

File: R/recipe_run.R (hunk 3 above)
Comment (r-ash) [reply R/recipe_run.R]: Yes I like that more - nicer to read than jumping up to the closure

File: R/recipe_commit.R (hunk 2 above)
Comment (r-ash) [reply R/recipe_commit.R]: Should be okay now we have the append mode when opening the file connection. This is tested too https://github.com/vimc/orderly/blob/16a58d1d01190b4d85edd727f022bde97257aac5/tests/testthat/test-run.R#L1010

File: R/util.R (hunk 4 above)
Comment (r-ash) [reply R/util.R]: > why not `<-`? and why is nolint required?

Just being absent minded. I've got rid of the nolint stuff too.

File: R/recipe_commit.R (hunk 2 above)
Comment (richfitz) [reply R/recipe_commit.R]: sorry, I'd meant to delete that comment once I realised we're appending!

**Pull request merged**